crates/bin/prove_block/src/main.rs

```diff
@@ -321,146 +387,113 @@ async fn main() -> Result<(), Box<dyn Error>> {
-        // really any overlap between contracts with storage updates (no nonce updates) and account
-        // contracts (nonces are updated), so this would be a corner case for now
-        let nonce = nonces.get(&contract_address).copied();
-
-        log::debug!("contract with storage diff, address: {} (nonce: {:?})", contract_address, nonce);
-        let contract_address_biguint = contract_address.to_biguint();
-
-        let address: ContractAddress = ContractAddress(PatriciaKey::try_from(felt_vm2api(contract_address)).unwrap());
-        let initial_contract_state = initial_state.get_contract_state(address)?;
-        let initial_tree = initial_contract_state.storage_commitment_tree.clone();
-
-        let contract_state = initial_contract_state
-            .update(
-                &mut initial_state.ffc,
-                &storage_diff_item.storage_entries.iter().map(|entry| (entry.key, entry.value)).collect(),
-                nonce,
-                address_to_class_hash.get(&contract_address).copied(),
-            )
-            .await?;
-        let updated_tree = contract_state.storage_commitment_tree.clone();
-
-        if storage_proof.class_commitment != contract_state.storage_commitment_tree.root.into() {
-            log::error!(
-                "expected class_commitment != computed class_commitment ({:?} != {:?})",
-                storage_proof.class_commitment,
-                contract_state.storage_commitment_tree.root.clone()
-            );
-        }
+    for (contract_address, _storage_proof) in storage_proofs {
+        let previous_storage_proof =
+            previous_storage_proofs.get(&contract_address).expect("failed to find previous storage proof");
 
-        updates.push((contract_address_biguint, contract_state.clone()));
-        contract_states.insert(storage_diff_item.address, contract_state);
+        let previous_tree = PatriciaTree { root: previous_storage_proof.state_commitment.into(), height: Height(251) };
 
-        let contract_storage =
-            OsSingleStarknetStorage::new(initial_tree, updated_tree, &[], initial_state.ffc.clone()).await?;
+        let contract_storage = ProverPerContractStorage::new(previous_block_id, contract_address, provider_url.clone());
         contract_storages.insert(contract_address, contract_storage);
-    }
 
-    // insert ContractState for any contract that received a nonce update but not a storage update
-    for nonce_update in state_update.state_diff.nonces {
-        if let std::collections::hash_map::Entry::Vacant(e) = contract_states.entry(nonce_update.contract_address) {
-            let mut contract_state = ContractState::empty(Height(251), &mut initial_state.ffc).await?;
+        log::debug!("contract address: {}", contract_address.to_hex_string());
+        let (class_hash, previous_nonce) = if [Felt252::ZERO, Felt252::ONE].contains(&contract_address) {
+            (Felt252::ZERO, Felt252::ZERO)
```

Comment:
Which case are you handling here? Can you add a comment? :)

Reply:
There are magic contracts at addresses `0x0` and `0x1` and I think this is handling these by setting class hash and nonce to 0, and that the OS expects this to be the case (@odesenfans wrote this so I can't be 100% sure).

I suggest we write a utility fn to detect if a given contract address is a magical one for clarity and extensibility. But, as you suggest, this could also use a comment in any case.

Reply:
@notlesh yes for now just adding the comment stating that there are some special contracts is ok :)